trieste/acquisition/function.py

```diff
@@ -509,6 +512,161 @@ def prepare_acquisition_function(
+            """
+            Generic Expected Improvement defined at Eq. 19 in :cite:`yang2019efficient`,
+            calculate the expected improvement on reference a, integrating from b
+            :param a: shape [num_cells, out_dim] expected improvement reference point
+            :param b: shape [num_cells, out_dim] integral upper bounds
+            :param mean: shape [..., out_dim]
+            :param std: shape [..., out_dim]
+            """
+            return std * normal.prob((b - mean) / std) + (mean - a) * (
+                1 - normal.cdf((b - mean) / std)
+            )
+
+        def nu(lb: TensorType, ub: TensorType, mean: TensorType, std: TensorType) -> TensorType:
+            """
+            A special expected improvement given fixed improvement,
+            defined at Eq. 25 in :cite:`yang2019efficient`
+            :param lb: shape [..., num_cells, out_dim] cell lower bound
+            :param ub: shape [..., num_cells, out_dim] cell upper bound
+            :param mean: shape [..., out_dim]
+            :param std: shape [..., out_dim]
+            """
+            return (ub - lb) * (1 - normal.cdf((ub - mean) / std))
+
+        def ehvi_based_on_partitioned_cell(psi_val: TensorType, nu_val: TensorType) -> TensorType:
+            r"""
+            Use psi and nu function value to calculate the ehvi based on cell i.
+            The original integration on each output dimensionality can be formulated as 2 type of
+            generic expected improvement corresponding to psi and nu respectively, hence
+            the integration through d dimension is a sum of combinations. Refer
+            Eq. 44, Eq. 45
+            .. math::
+                \int_{y_1=l_1}^{\infty}\int_{y_2=l_2}^{\infty},...,\int_{y_d=l_d}^{\infty}
+                \lambda_d[S_d^{(i)}\cap\Delta(y)]\cdot\xi_{\mu,\sigma}(y)dy
+                = \Sigma_{j=1}^{2^d}(\prod_{k=1}^d \omega(.,.,C_k^{j})
+            Which can be reformulated as combinations of Psi and nu
+            :param psi_val shape [..., num_cells, out_dim]
+            :param nu_val shape [..., num_cells, out_dim]
+            """
+
+            stacked_factors = tf.concat(
+                [tf.expand_dims(psi_val, -2), tf.expand_dims(nu_val, -2)], axis=-2
+            )  # Take the cross product of psi_diff and nu across all outcomes
+            # [..., num_cells, 2(operation_num, refer Eq. 45), num_obj]
+
+            factor_combinations = tf.linalg.diag_part(
+                tf.gather(stacked_factors, cross_index, axis=-2)
+            )  # [..., num_cells, 2^d, 2(operation_num), num_obj]
+
+            return tf.reduce_sum(tf.reduce_prod(factor_combinations, axis=-1), axis=-1)
+
+        candidate_mean, candidate_var = model.predict(tf.squeeze(x, -2))
+        candidate_std = tf.sqrt(candidate_var)
+
+        neg_candidate_mean = -tf.expand_dims(candidate_mean, 1)  # [..., 1, out_dim]
+        candidate_std = tf.expand_dims(candidate_std, 1)  # [..., 1, out_dim]
```

Comment:
Could the new function not include (almost) all the lines below here? i.e. dealing with all the upper bound stuff?